pkg/sentry/server/ca/ca.go

```diff
@@ -0,0 +1,231 @@
+	TrustAnchors() []byte
+}
+
+// store is the interface for the trust bundle backend store.
+type store interface {
+	store(context.Context, caBundle) error
+	get(context.Context) (caBundle, bool, error)
+}
+
+// ca is the implementation of the CA Interface.
+type ca struct {
+	bundle caBundle
+	config config.Config
+}
+
+// caBundle is the bundle of certificates and keys used by the CA.
+type caBundle struct {
+	trustAnchors []byte
+	issChainPEM  []byte
+	issKeyPEM    []byte
+	issChain     []*x509.Certificate
+	issKey       any
+}
+
+func New(ctx context.Context, conf config.Config) (Interface, error) {
+	var castore store
+	if config.IsKubernetesHosted() {
+		log.Info("using kubernetes secret store for trust bundle storage")
+
+		restConfig, err := rest.InClusterConfig()
+		if err != nil {
+			return nil, err
+		}
+		client, err := kubernetes.NewForConfig(restConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		castore = &kube{
+			config:    conf,
+			namespace: security.CurrentNamespace(),
+			client:    client,
+		}
+	} else {
+		log.Info("using local file system for trust bundle storage")
+		castore = &selfhosted{config: conf}
+	}
+
+	bundle, ok, err := castore.get(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		log.Info("root and issuer certs not found: generating self signed CA")
+
+		bundle, err = generateCABundle(conf)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Info("root and issuer certs generated")
+
+		if err := castore.store(ctx, bundle); err != nil {
+			return nil, err
+		}
+
+		log.Info("self signed certs generated and persisted successfully")
+		monitoring.IssuerCertChanged()
+		monitoring.IssuerCertExpiry(bundle.issChain[0].NotAfter)
+	} else {
+		log.Info("root and issuer certs found: using existing certs")
+	}
+
+	return &ca{
+		bundle: bundle,
+		config: conf,
+	}, nil
+}
+
+func (c *ca) SignIdentity(ctx context.Context, req *SignRequest) ([]*x509.Certificate, error) {
+	spiffeID, err := (spiffe.Parsed{
+		TrustDomain: req.TrustDomain,
+		Namespace:   req.Namespace,
+		AppID:       req.AppID,
+	}).ToID()
+	if err != nil {
+		return nil, err
+	}
+
+	tmpl, err := generateWorkloadCert(req.SignatureAlgorithm, c.config.WorkloadCertTTL, c.config.AllowedClockSkew, spiffeID)
+	if err != nil {
+		return nil, err
+	}
+	tmpl.DNSNames = append(tmpl.DNSNames, req.DNS...)
+
+	certDER, err := x509.CreateCertificate(rand.Reader, tmpl, c.bundle.issChain[0], req.PublicKey, c.bundle.issKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return nil, err
+	}
+
+	return append([]*x509.Certificate{cert}, c.bundle.issChain...), nil
+}
+
+// TODO: Remove this method in v1.12 since it is not used any more.
+func (c *ca) TrustAnchors() []byte {
+	return c.bundle.trustAnchors
+}
+
+func generateCABundle(conf config.Config) (caBundle, error) {
+	rootKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```

Comment:
More like a wishlist item (for a separate PR), but would be great to have support for other signing algorithms like good ol' RSA or EdDSA (Ed25519) too…